Change all 'hashlib.md5' to 'hashlib.sha1'

g19fanatic opened the pull request:
Tied to issue #31800

g19fanatic committed:
Change all 'hashlib.md5' to 'hashlib.sha1'

g19fanatic commented:
So what can I do to get this through? I've already made hot changes to my pip install of tensorflow (essentially applying this patch) and I've been able to train models and get them to predict and evaluate without much issues..

mihaimaruseac reviewed (approved):
Thank you for the PR.

mihaimaruseac commented on tensorflow/python/keras/utils/data_utils.py:
```diff
@@ -293,7 +293,7 @@ def _hash_file(fpath, algorithm='sha256', chunk_size=65535):
     hasher = hashlib.sha256()
   else:
-    hasher = hashlib.md5()
+    hasher = hashlib.sha1()
```
From @fchollet 

> The behavior of `_hash_file` needs to stay backwards compatible since the output is used to match cached files that are already written (i.e. it interacts with the real world, it is not an internal implementation detail). Unless there are critical concerns, we should revert this line.

mihaimaruseac reviewed (changes requested):
Seems this is going to break backwards compatibility if done this way.

We need more care to ensure no op is broken by this change

caisq reviewed (dismissed):
LGTM for the changes in the /debug directories.

rthadur commented:
@g19fanatic can you please resolve merge conflicts and check build failures ?

g19fanatic commented:
Will do!

On Fri, Oct 25, 2019, 2:37 PM Rajeshwar Reddy T <notifications@github.com>
wrote:

> @g19fanatic <https://github.com/g19fanatic> can you please resolve merge
> conflicts and check build failures ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/31803?email_source=notifications&email_token=AAEAZU3LTZZKGFPYYJLHNC3QQM4EZA5CNFSM4INZOWZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECJFYYA#issuecomment-546462816>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEAZU65K6XOIVVYYDTTO23QQM4EZANCNFSM4INZOWZA>
> .
>

mihaimaruseac commented:
Please also consider the compatibility issue from my review

g19fanatic committed:
revert hasher back to md5 for _hash_file

only 2 locations 'in' tensorflow (besides tests) use md5 (AFAIK)

./python/keras/datasets/imdb.py-145-  origin_folder = 'https://storage.googleapis.com/tensorflow/tf-keras-datasets/'
./python/keras/datasets/reuters.py-129-  origin_folder = 'https://storage.googleapis.com/tensorflow/tf-keras-datasets/'

g19fanatic commented:
@mihaimaruseac  & @fchollet 

I reverted back that one specific change in data_utils.

A quick search shows 2 locations in the codebase (besides tests testing the md5 hashing) that utilizes md5 for the file_hashing. This shouldn't break anyone who is still using these datasets (namely, imdb and reuters) on a non-FIPS compliant machine.

Anyone who is specifically caring about their hashing function would be broken under a FIPS compliant OS but eh? Don't use depreciated hashing functions in new stuff? If you're trying to use the imdb or reuters datasets, I imagine the codebase could be updated to have the proper hash for the files...

This PR doesn't break my project and I don't specifically set any sort of hashing function and the default goes to sha265 anyways.

mihaimaruseac commented:
The specific function has this signature `def _hash_file(fpath, algorithm='sha256', chunk_size=65535):` It already has different behavior based on the value of `algorithm`. I would suggest to add the change back under an `elif` to make the function support more that just `sha256` and `md5` hashes.

g19fanatic commented:
> 
> 
> The specific function has this signature `def _hash_file(fpath, algorithm='sha256', chunk_size=65535):` It already has different behavior based on the value of `algorithm`. I would suggest to add the change back under an `elif` to make the function support more that just `sha256` and `md5` hashes.

This is specifically what my last commit does...
[Adds back the change](https://github.com/tensorflow/tensorflow/pull/31803/commits/aa170db3af310dcb8545d6ec7476163014d46808#diff-ba2b45d565934dc5120efeef5457ea2eR320) under the elif under the `_hash_file` function.

g19fanatic commented:
> 
> 
> There is no `elif` in that patch.

I see what you're saying. You'd like to support MORE than the originally supported sha265 and md5.

That would require many more locations to be updated (and documentation). Honestly, out of scope of what I was trying to do with this issue and PR (namely make tensorflow work for FIPS compliant machines). 

My provided PR has reverted the one `else:` back to where it was before as was described as a potential issue with my previous patch.

g19fanatic commented:
> 
> 
> I was only suggesting to add the sha1 to the specific file. But let me investigate more to see which other functions would need this and I'll come back here.

As far as I can tell,   functions `validate_file` and `get_file` are in the call chain to `_hash_file`.

besides tests, `_hash_file` is only used in `validate_file` and a tutorial file<br>
`validate_file` is only used in `get_file` <br>
`get_file` is only used in 
```
./lite/g3doc/models/style_transfer/overview.ipynb
./lite/g3doc/performance/post_training_quant.ipynb
./python/keras/datasets/boston_housing.py
./python/keras/datasets/cifar10.py
./python/keras/datasets/cifar100.py
./python/keras/datasets/fashion_mnist.py
./python/keras/datasets/imdb.py
./python/keras/datasets/mnist.py
./python/keras/datasets/reuters.py
./python/keras/utils/data_utils.py
./python/keras/utils/data_utils_test.py
```

and only imdb and reuters use md5 as the hashing algo.. every other one uses default options...

leaving it as is should be okay...

mihaimaruseac commented:
There are also issue with Google internal usages. That's why it needs more look.

gbaned commented:
@mihaimaruseac Any update on this PR, please. Thanks!

mihaimaruseac commented:
I will try to import this manually over the weekend. Thanks for the ping

gbaned commented:
@g19fanatic Can you please resolve conflicts? Thanks!

mihaimaruseac commented:
I am in the process of importing this manually (though changing to sha256 instead of sha1).

tqa236 commented:
Hello all, it's been a while, any news on this PR?

gbaned commented:
@g19fanatic Can you please resolve conflicts? Thanks!

gbaned commented:
@mihaimaruseac, Any update on this PR? Please. Thanks!

g19fanatic commented:
I too am waiting on this fix. Deploying any code requires a patch to be applied for FIPS to not complain

mihaimaruseac commented:
I am starting an internal process to change these. We should have this done by the 2.5 release early next year.
